packages/remesh/src/modules/list.ts

```diff
@@ -132,32 +152,50 @@ export const ListModule = <T extends SerializableObject>(
     },
   })
 
+  const updateList = (oldList: T[], newItem: T) => {
+    const key = options.key(newItem)
+    const newList = [] as T[]
+
+    let isChanged = false
+
+    for (const item of oldList) {
+      if (options.key(item) === key) {
+        newList.push(newItem)
+        isChanged = true
+      } else {
+        newList.push(item)
+      }
+    }
+
+    if (isChanged) {
+      return newList
+    }
+    return null
+  }
+
   const UpdateItemCommand = domain.command({
     name: `${options.name}.UpdateItemCommand`,
     impl: ({ get }, newItem: T) => {
       const oldList = get(ItemListState())
-      const key = options.key(newItem)
-      const newList = [] as T[]
-
-      let isChanged = false
-
-      for (const item of oldList) {
-        if (options.key(item) === key) {
-          newList.push(newItem)
-          isChanged = true
-        } else {
-          newList.push(item)
-        }
-      }
+      const newList = updateList(oldList, newItem)
 
-      if (isChanged) {
+      if (newList) {
         return ItemListState().new(newList)
       }
 
       return null
     },
   })
 
+  const UpsertItemCommand = domain.command({
+    name: `${options.name}.UpsertItemCommand`,
+    impl: ({ get }, newItem: T) => {
+      const oldList = get(ItemListState())
+      const newList = updateList(oldList, newItem)
+      return ItemListState().new(newList ?? oldList.concat(newItem))
```

Comment:
`concat` is not a good way to add item in list with generic item type, see below

```ts
[].concat(1) // [1]
[].concat([1]) // [1]
```

use `list spread operator` will be better

```ts
[...oldList, newItem]
```

Reply:
Since `oldList` will be type `T[]`, and the type of `newItem` is `T`, it should be ok here.